src/main/java/org/jenkinsci/plugins/workflow/steps/ErrorInfoStep.java

```diff
@@ -0,0 +1,193 @@
+ */
+public class ErrorInfoStep extends AbstractStepImpl {
+
+    public final Throwable error;
+
+    @DataBoundConstructor public ErrorInfoStep(Throwable error) {
+        this.error = error;
+    }
+
+    public static class Execution extends AbstractSynchronousStepExecution<ErrorInfo> {
+
+        private static final long serialVersionUID = 1;
+        @Inject private transient ErrorInfoStep step;
+        @StepContextParameter private transient FlowExecution execution;
+
+        @Override protected ErrorInfo run() throws Exception {
+            return new ErrorInfo(step.error, execution);
+        }
+
+    }
+
+    @Extension public static class DescriptorImpl extends AbstractStepDescriptorImpl {
+
+        public DescriptorImpl() {
+            super(Execution.class);
+        }
+
+        @Override public String getFunctionName() {
+            return "errorInfo";
+        }
+
+        @Override public String getDisplayName() {
+            return "Calculate information about an error";
+        }
+
+        // TODO blank config.jelly
+
+    }
+
+    public static class ErrorInfo implements Serializable {
+
+        private static final long serialVersionUID = 1;
+        private final Throwable error;
+        private transient FlowExecution execution;
+        private final FlowExecutionOwner executionOwner;
+
+        ErrorInfo(Throwable error, FlowExecution execution) {
+            this.error = error;
+            this.execution = execution;
+            executionOwner = execution.getOwner();
+        }
+
+        private FlowExecution getExecution() throws IOException {
+            if (execution == null) {
+                execution = executionOwner.get();
+            }
+            return execution;
+        }
+
+        /**
+         * Finds a node which threw this exception or one of its causes.
+         * Note that {@link Throwable#equals} is just pointer equality,
+         * which we cannot use since we may be loading deserialized exceptions,
+         * so we compare by stack trace instead.
+         */
+        private @CheckForNull FlowNode getNode() throws IOException {
+            final Set<String> stackTraces = new HashSet<>();
+            for (Throwable t = error; t != null; t = t.getCause()) {
+                stackTraces.add(Functions.printThrowable(t));
+            }
+            Predicate<FlowNode> threwException = new Predicate<FlowNode>() {
+                @Override
+                public boolean apply(FlowNode input) {
+                    if (input instanceof BlockEndNode) {
+                        return false;
+                    }
+                    ErrorAction a = input.getAction(ErrorAction.class);
+                    return (a != null && stackTraces.contains(Functions.printThrowable(a.getError())));
+                }
+            };
+            return new FlowScanner.DepthFirstScanner().findFirstMatch(getExecution(), threwException);
```

Comment:
Performance is not a significant concern for this PR, because this is code run just once per step execution, typically once per build, and from the CPS VM thread. Obviously if you have some system for indexing nodes using specific action classes, that would be helpful as you could go straight to those with `ErrorAction`.

I did not follow the discussion about `ForkScanner` here. At any rate, for purposes of this PR we are looking for any `FlowNode` in the build with an `ErrorAction`, so long as it is not a `BlockEndNode`. Possibly we really want to check also `BlockEndNode`s, to handle errors thrown only during cleanup from build wrappers and the like, but give them a lower priority. The issue is that `ErrorAction` gets attached to end nodes of blocks which are exited by an uncaught exception, so we want to find the “first” node with the matching action: i.e., giving preference to the nested node. So it would also suffice to do a walk forward in time, rather than backward as I presume `DepthFirstScanner` does. Seems there is no such scanner yet?